### PR TITLE
Expand aggregation function options for all supported fields

### LINKED ISSUE #12248: Missing aggregate functions after joining tables (1st table has limited aggregate possibilities
**Describe the bug**
When you have a small table with limited aggregate possibilities.
image

And you want to join it with a table that has more aggregate possibilities, e.g. sum of `duration`.
image

Then you cannot achieve this in the query builder because it seems to focus on the aggregate capabilities of the first table.

**To Reproduce**
Joining the two table in the query builder and you will that I cannot apply the aggregate function the column duration from the second table.
image

**Expected behavior**
I expected to be able to apply the sum aggregate function on the duration after the joining the two tables

**Screenshots**

**Information about your Metabase Installation:**
```
{
  "browser-info": {
    "language": "en",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_4) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.1 Safari/605.1.15",
    "vendor": "Apple Computer, Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "1.8.0_242-heroku-b08",
    "java.vendor": "Oracle Corporation",
    "java.vendor.url": "http://java.oracle.com/",
    "java.version": "1.8.0_242-heroku",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "25.242-b08",
    "os.name": "Linux",
    "os.version": "4.4.0-1062-aws",
    "user.language": "en",
    "user.timezone": "Etc/UTC"
  },
  "metabase-info": {
    "databases": [
      "postgres",
      "googleanalytics"
    ],
    "hosting-env": "heroku",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "11.6 (Ubuntu 11.6-1.pgdg16.04+1)"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.2.8"
      }
    },
    "run-mode": "prod",
    "version": {
      "tag": "v0.35.0",
      "date": "2020-03-31",
      "branch": "master",
      "hash": "3f96c47"
    },
    "settings": {
      "report-timezone": "Europe/Berlin"
    }
  }
}
```
**Severity**
I was possible to solve it in another way. But was wondering this is. I tried changing the settings of the tables. But I could not manage to get it to work.

### PR DESCRIPTION
**Pull request by deniskaber**
Closes https://github.com/metabase/metabase/issues/12248

### Description

It is not possible to add an aggregation for fields from implicitly joined tables.

### How to verify

Use case 1 - https://github.com/metabase/metabase/issues/12248#issuecomment-730270741

Use case 2 -

1. Use the same setup as in use case 1 in the admin site.
2. Start new Question -> Reviews
3. Add explicit join with Products
4. Try to add an aggregation based on `Products.Price`

### Demo

https://user-images.githubusercontent.com/7983744/228362964-5ee95e18-cacd-464f-80a3-c066aa74b22a.mov

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/29623)

### CODE REVIEW

**Review by ranquild — APPROVED** [review]

**Review by WiNloSt — CHANGES_REQUESTED** [review]
@ranquild ~I'm not sure what's the behavior should be for models, but it makes sense to me that we should only take models' visible fields into account.~ Looks like what I've seen wasn't an invisible fields on a model, but rather a field from the foreign table

The way I set this up is
1. Created a model based on any table, but only have foreign key fields
2. Test on `master` that in query builder, we can only aggregated on the visible fields e.g. in this case there's no numeric fields, so no aggregations for numeric values should be presented in the options.
3. Test on this PR that in the query builder, even though there's no numeric fields visible, there's still aggregation options for invisible fields, and it even works when trying to visualize the query.

~I'm not sure (3) is the desire behavior, but that's not what we currently do on `master` (c9542cef94f8b73b333b3bf219e68a5712433b34 for reference)~, I think this is the correct behavior. Normally, on an unaggregated table, we can choose to aggregation fields from a foreign table.

## The model that doesn't have any numeric fields
image
image

## This PR can aggregate ~invisible fields on models~. I was wrong, this showed the field from a foreign table
image
image

## `master` at c9542cef94f8b73b333b3bf219e68a5712433b34 don't show **Average of ...** option

image

**Review by WiNloSt — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29623)* status: 0 of 6 files reviewed, 3 unresolved discussions (waiting on @deniskaber)

___
*[`frontend/src/metabase-lib/operators/utils/index.unit.spec.js` line 151 at r2](https://reviewable.io/reviews/metabase/metabase/29623#-NRmc4EZDURWh8i_ZhcR:-NRmc4EZDURWh8i_ZhcS:b-numkbn) ([raw file](https://github.com/metabase/metabase/blob/42abfbcbef6f7e888d481a54768eb6508330ee05/frontend/src/metabase-lib/operators/utils/index.unit.spec.js#L151)):*
> ```JavaScript
> 
>     it("returns nothing without DB features", () => {
>       const db = getDatabase([]);
> ```

Please refrain from using abbreviated variable names. I notice `Table.db` exists, but with Metabase Lib v2 it should be gone.

Some abbreviations feel universal like `db` in this case, but it's worth noting that some people in Metabase have used DB for dashboard which threw me off once.

In general abbreviating names creates bad practice since some abbreviations are not as obvious as others, and it might feel like second nature to writers, but not to all readers.
___
*[`frontend/src/metabase-lib/queries/StructuredQuery.ts` line 656 at r2](https://reviewable.io/reviews/metabase/metabase/29623#-NRmbLj7376GGM9kYuuJ:-NRmbLj7376GGM9kYuuK:blh5wbi) ([raw file](https://github.com/metabase/metabase/blob/42abfbcbef6f7e888d481a54768eb6508330ee05/frontend/src/metabase-lib/queries/StructuredQuery.ts#L656)):*
> ```XML
>       const fieldOptions = this.fieldOptions()
>         .all()
>         .map(dimension => dimension.field?.())
> ```

I don't think dimension.field (I'm talking about the `field` function itself, not the return value from it) could be null. I check all 4 classes that extend `Dimension` class, and they all have `field` function.
1. `FieldDimension`
2. `ExpressionDimension`
3. `AggregationDimension`
4. `TemplateTagDimension` (which extends `FieldDimension`)
___
*[`frontend/src/metabase-lib/queries/StructuredQuery.ts` line 657 at r2](https://reviewable.io/reviews/metabase/metabase/29623#-NRmbWQ2EgwRde8vzzRM:-NRmbWQ2EgwRde8vzzRN:b-q5dwd0) ([raw file](https://github.com/metabase/metabase/blob/42abfbcbef6f7e888d481a54768eb6508330ee05/frontend/src/metabase-lib/queries/StructuredQuery.ts#L657)):*
> ```XML
>         .all()
>         .map(dimension => dimension.field?.())
>         .filter(field => field != null);
> ```

We have a semantic function for this https://github.com/metabase/metabase/blob/d27156538c9046a5f7631bd7e6847b82ccfd736b/frontend/src/metabase/core/utils/types/types.ts#L1

**Review by WiNloSt — APPROVED** [review]
Wasn't meant to request changes to block the PR from merging, but I've added some small comments that would appreciate you modifying before merging.

**Pull request merged**
